core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala

```diff
@@ -351,58 +327,37 @@ object LoadBalancerService {
      * @param invokerBusyThreshold defines when an invoker is considered overloaded
      * @param hash stable identifier of the entity to be scheduled
      * @return an invoker to schedule to or None of no invoker is available
      */
-    def schedule[A](
-        availableInvokers: Seq[A],
-        activationsPerInvoker: collection.Map[A, Int],
+    def schedule(
+        invokers: Seq[(InstanceId, InvokerState, Int)],
         invokerBusyThreshold: Int,
-        hash: Int): Option[A] = {
+        hash: Int): Option[InstanceId] = {
 
-        val numInvokers = availableInvokers.length
+        val numInvokers = invokers.size
         if (numInvokers > 0) {
             val homeInvoker = hash % numInvokers
-
             val stepSizes = LoadBalancerService.pairwiseCoprimeNumbersUntil(numInvokers)
             val step = stepSizes(hash % stepSizes.size)
 
-            @tailrec
-            def search(targetInvoker: Int, iteration: Int = 1): A = {
-                // map the computed index to the actual invoker index
-                val invokerName = availableInvokers(targetInvoker)
-
-                // send the request to the target invoker if it has capacity...
-                if (activationsPerInvoker.get(invokerName).getOrElse(0) < invokerBusyThreshold * iteration) {
-                    invokerName
-                } else {
-                    // ... otherwise look for a less loaded invoker by stepping through a pre-computed
-                    // list of invokers; there are two possible outcomes:
-                    // 1. the search lands on a new invoker that has capacity, choose it
-                    // 2. walked through the entire list and found no better invoker than the
-                    //    "home invoker", force the home invoker
-                    val newTarget = (targetInvoker + step) % numInvokers
-                    if (newTarget == homeInvoker) {
-                        if (iteration < 3) {
-                            search(newTarget, iteration + 1)
-                        } else {
-                            availableInvokers(homeInvoker)
-                        }
-                    } else {
-                        search(newTarget, iteration)
-                    }
-                }
-            }
-
-            Some(search(homeInvoker))
-        } else {
-            None
-        }
+            val invokerProgression = Stream.from(0)
+                .take(numInvokers)
+                .map(i => (homeInvoker + i * step) % numInvokers)
+                .map(invokers)
+                .filter(_._2 == Healthy)
+
+            invokerProgression.find(_._3 < invokerBusyThreshold)
+                .orElse(invokerProgression.find(_._3 < invokerBusyThreshold * 2))
+                .orElse(invokerProgression.find(_._3 < invokerBusyThreshold * 3))
+                .orElse(invokerProgression.headOption)
+                .map(_._1)
+        } else None
```

Comment:
the hard part about reading this was figure out how to destructure the tuple... almost tempted for a case class.